[lld][ELF][X86] Add missing X86_64_TPOFF64 case in switches

nmosier opened the pull request:
Close #77201. When linking code with a R_X86_64_TPOFF64 relocation, LLD exits with an 'unknown reloaction' error message due to two missing cases in relocation switch statements. This patch adds in those cases so that LLD successfully links code R_X86_64_TPOFF64 relocations.

nmosier committed:
[lld][ELF][X86] Add missing X86_64_TPOFF64 case in switches

Close #77201. When linking code with a R_X86_64_TPOFF64 relocation,
LLD exits with an 'unknown relocation' error message due to two missing
cases in relocation switch statements. This patch adds in those cases
so that LLD successfully links code R_X86_64_TPOFF64 relocations.

MaskRay commented:
Thanks for the patch. https://maskray.me/blog/2021-08-08-toolchain-testing#the-test-checks-too-little
x86-64 TLS tests are not very well organized, but you can add a test to `x86-64-tls-pie.s`  and add `llvm-objdump -d --no-show-raw-insn` to check the disassembly.

nmosier commented:
Thanks for the comments. I updated the x86-64-tls-pie.s test as you suggested and removed the new one I previously had added. I also updated the commit message as suggested.

MaskRay commented on lld/test/ELF/x86-64-tls-pie.s:
```diff
@@ -1,14 +1,23 @@
-# Bug 27174: R_X86_64_TPOFF32 and R_X86_64_GOTTPOFF relocations should
-# be eliminated when building a PIE executable, as the static TLS layout
-# is fixed.
+# Bug 27174: R_X86_64_TPOFF32, R_X86_86_TPOFF64, and R_X86_64_GOTTPOFF
```
The comment should be left as is as lld didn't support `R_X86_86_TPOFF64` as that time.

You can add a comment that this additionally tests `R_X86_86_TPOFF64`

MaskRay commented on lld/test/ELF/x86-64-tls-pie.s:
```diff
@@ -1,14 +1,23 @@
+# DIS-NEXT:                movq    %fs:0, %rax
+# DIS-NEXT:                movl    $3, (%rax)
+# DIS-NEXT:                movq    %fs:0, %rdx
+# DIS-NEXT:                movq    (%rip), %rcx            # 0x23 <_start+0x23>
```
Drop `# 0x23 <_start+0x23>` so that the test is less sensitive to exact address/offset values.

MaskRay commented:
> Thanks for the comments. I updated the x86-64-tls-pie.s test as you suggested and removed the new one I previously had added. I also updated the commit message as suggested.

Thanks. You can update the first comment, which will actually be used when a user with write permission clicks "Squash and merge"... The commit message is actually ignored.

nmosier commented:
> Drop `# 0x23 <_start+0x23>`
Done.

> You can add a comment that this additionally tests R_X86_86_TPOFF64
Done.

> Thanks. You can update the first comment,
Updated.

> It also helps to state which producer (GCC?) creates `R_X86_64_TPOFF64`. A psABI may define many not-so-useful relocation types and we may not support all of them.

Actually, it's LLVM/clang. I have submitted a concurrent patch for allowing LLVM to emit R_X86_64_TPOFF64 relocations, which a project I'm working on requires.

MaskRay commented:
Thanks for the update.

> Actually, it's LLVM/clang. I have submitted a concurrent patch for allowing LLVM to emit R_X86_64_TPOFF64 relocations, which a project I'm working on requires.

This is interesting. R_X86_64_TPOFF32 supports offsets up to -2**31, which is much more than sufficient.
Extremely large TLS offsets likely do not work on many libc/rtld implementations.
